File: Sources/LibraryModel/Volume.swift
//
// Volume.swift
// Copyright (c) 2021 Paul Schifferer.
//

import Fluent
import Vapor


public final class Volume : Model {
    public static let schema = Volume.v20210620.schemaName

    @ID
    public var id : UUID?

    @Timestamp(key: Volume.v20210620.createdAt, on: .create)
    public var createdAt : Date?

    @Timestamp(key: Volume.v20210620.updatedAt, on: .update)
    public var updatedAt : Date?

    @Field(key: Volume.v20210620.name)
    public var name : String

    @Field(key: Volume.v20210620.slug)
    public var slug : String

    @Field(key: Volume.v20210620.isbn)
    public var isbn : String

    @Siblings(through: VolumeAuthorPivot.self, from: \.$volume, to: \.$author)
    public var authors : [Author]

    @Siblings(through: VolumeStudioPivot.self, from: \.$volume, to: \.$studio)
    public var studios : [Studio]

    @Siblings(through: VolumePublisherPivot.self, from: \.$volume, to: \.$publisher)
    public var publishers : [Publisher]

    @Siblings(through: VolumeTagPivot.self, from: \.$volume, to: \.$tag)
    public var tags : [Tag]

    @Parent(key: Volume.v20210625.systemId)
    public var system : LibraryModel.System

    @Children(for: \.$volume)
    public var reviews : [Review]

    @Children(for: \.$volume)
    public var properties : [VolumeProperty]

    @Timestamp(key: Volume.v20210620.deletedAt, on: .delete)
    public var deletedAt : Date?

    public init() {
    }

    public init(id : UUID? = nil, name : String, systemId : LibraryModel.System.IDValue) {
        self.id = id
        self.name = name
        self.$system.id = systemId
    }

}

extension Volume : Content {}

File: src/sweetrpg_library_model/volume.py
# -*- coding: utf-8 -*-
__author__ = "Paul Schifferer <paul@schifferers.net>"
"""
"""

from datetime import datetime


class Volume(object):
      """
      """
      __tablename__ = 'volumes'

      def __init__(self, *args, **kwargs):
            """
            """
            now = datetime.utcnow().isoformat()
            self.id = kwargs.get('_id') or kwargs.get('id')
            self.name = kwargs['name']
            self.created_at = kwargs.get('created_at', now)
            self.updated_at = kwargs.get('updated_at', now)
            self.deleted_at = kwargs.get('deleted_at')
            self.system = kwargs['system']
            self.slug = kwargs['slug']
            self.isbn = kwargs.get('isbn')
            self.authors = kwargs.get('authors', [])
            self.publishers = kwargs.get('publishers', [])
            self.studios = kwargs.get('studios', [])
            self.reviews = kwargs.get('reviews', [])

      def __repr__(self):
            return f"""<Volume(id={self.id!r},
                        name={self.name!s}, slug={self.slug!s},
                        system={self.system!s}, isbn={self.isbn!s},
                        authors={self.authors!r},
                        publishers={self.publishers!r},
                        studios={self.studios!r},
                        reviews={self.reviews!r},
                        created_at={self.created_at!r}, updated_at={self.updated_at!r}, deleted_at={self.deleted_at!r})>
                        """
